pyo3-macros-backend/src/method.rs

```diff
@@ -735,3 +742,5 @@ fn parse_method_attributes(
 }
 
 const IMPL_TRAIT_ERR: &str = "Python functions cannot have `impl Trait` arguments";
+const RECEIVER_BY_VALUE_ERR: &str = "PyO3 does not support `self` as a receiver, because it cannot move `self` out of the Python interpreter.
```

Comment:
I do not really have a good proposal but since this is primarily about the error message, I wanted to say that "moving out of the Python interpreter" confused me a bit during the first reading.

AFAIU, the main issue is that all Python objects have shared ownership (similar to e.g. `Rc`), so we cannot leave a deinitialized object behind? (I think we could theoretically implement fallibly moving out of objects if their reference count is one, but I suspect that this basically never happens in Python code.)

So maybe we could word this in such a manner? Maybe something like "it cannot move from `self` as other Python references to it might still exist"?

Reply:
I've always thought of it as "the interpreter owns the object, you just own a reference to it". So to me it makes most sense to say that you can't move it out.

That said I think it doesn't really matter for the user to understand it perfectly if one of the suggestions will do.

Reply:
> "the interpreter owns the object, you just own a reference to it"

That sounds a bit arbitrary as the interpreter will not hold on to objects just for the sake of it, only because other Python code has references to it (which should apply even to tracing JIT like PyPy).

If only the interpreter would hold on to those objects, it would probably also not be a problem since the only completely generic thing a compiler can do to any object is to destroy/deallocate it so we would just need to leave a `PyCell` that can be destroyed.

(Writing that, we could probably support moving out of `self` if `PyCell` were adjust to support only `2^64 - 3` instead of `2^64 - 2` shared borrows thereby creating another niche that can be used for moved from values turning it into something like a `RefCell<Option<T>>`, but the ergonomics of all subsequent method invocations not explicitly taking `slf: &PyCell<T>` panicking would probably be less than nice.)

Reply:
https://github.com/PyO3/pyo3/issues/2225 would be a better place to discuss whether we want to support something like that.

> "it cannot move from self as other Python references to it might still exist"?

I believe the reference count is incremented when the method is called (?) so there will always be other references to it?

Reply:
> I believe the reference count is incremented when the method is called (?) so there will always be other references to it?

Yes, function arguments always seem to have an elevated reference count with CPython. This is what I meant with

> I think we could theoretically implement fallibly moving out of objects if their reference count is one, but I suspect that this basically never happens in Python code.

So maybe we could just say that "Python objects always have shared ownership"? (Mainly because "shared ownership" is term used in the official Rust documentation so people might have read it before.)

Reply:
> https://github.com/PyO3/pyo3/issues/2225 would be a better place to discuss whether we want to support something like that.

Agreed. Moved the comment there.

Reply:
@adamreichold what do you think about "Python objects are shared, so 'self' cannot be moved out of the Python interpreter."

Reply:
Agreed. I would find "Python object always have shared ownership, so `self` cannot be moved out of the Python interpeter" even nicer but I'd be happy as you propose.

Reply:
I have to agree that "Python interpreter" as the owner sounds strange here.